1-a

In [10]:
import matplotlib.pyplot as plt #import the matplotlib
%matplotlib inline 
import numpy as np
import pandas as pd

In [5]:
data = pd.read_csv('data/imports-85-01.csv', sep = ",")

In [3]:
missing_values = ["?"]
data1 = pd.read_csv("data/imports-85-01.csv", na_values = missing_values)

In [7]:
data.head()

make fuel-type  length  width  height horsepower  city-mpg  \
0  alfa-romero       gas   168.8   64.1    48.8        111        21   
1  alfa-romero       gas   168.8   64.1    48.8        111        21   
2  alfa-romero       gas   171.2   65.5    52.4        154        19   
3         audi       gas   176.6   66.2    54.3        102        24   
4         audi       gas   176.6   66.4    54.3        115        18   

   highway-mpg  price  
0           27  13495  
1           27  16500  
2           26      ?  
3           30      ?  
4           22  17450

In [8]:
data1.head()

make fuel-type  length  width  height  horsepower  city-mpg  \
0  alfa-romero       gas   168.8   64.1    48.8       111.0        21   
1  alfa-romero       gas   168.8   64.1    48.8       111.0        21   
2  alfa-romero       gas   171.2   65.5    52.4       154.0        19   
3         audi       gas   176.6   66.2    54.3       102.0        24   
4         audi       gas   176.6   66.4    54.3       115.0        18   

   highway-mpg    price  
0           27  13495.0  
1           27  16500.0  
2           26      NaN  
3           30      NaN  
4           22  17450.0

In [9]:
print(data1.isnull().sum().sum())

8


8 missing values in data

1-b

In [17]:
data2 = data1.dropna()
data2

make fuel-type  length  width  height  horsepower  city-mpg  \
0    alfa-romero       gas   168.8   64.1    48.8       111.0        21   
1    alfa-romero       gas   168.8   64.1    48.8       111.0        21   
4           audi       gas   176.6   66.4    54.3       115.0        18   
5           audi       gas   177.3   66.3    53.1       110.0        19   
6           audi       gas   192.7   71.4    55.7       110.0        19   
..           ...       ...     ...    ...     ...         ...       ...   
200        volvo       gas   188.8   68.9    55.5       114.0        23   
201        volvo       gas   188.8   68.8    55.5       160.0        19   
202        volvo       gas   188.8   68.9    55.5       134.0        18   
203        volvo    diesel   188.8   68.9    55.5       106.0        26   
204        volvo       gas   188.8   68.9    55.5       114.0        19   

     highway-mpg    price  
0             27  13495.0  
1             27  16500.0  
4             22  17450.0  
5             25  15250.0  
6             25  17710.0  
..           ...      ...  
200           28  16845.0  
201           25  19045.0  
202           23  21485.0  
203           27  22470.0  
204           25  22625.0  

[197 rows x 9 columns]

I drop(delete) the missing value

1-c

In [22]:
data2.describe(include = "all")

make fuel-type      length       width      height  horsepower  \
count      197       197  197.000000  197.000000  197.000000  197.000000   
unique      21         2         NaN         NaN         NaN         NaN   
top     toyota       gas         NaN         NaN         NaN         NaN   
freq        32       177         NaN         NaN         NaN         NaN   
mean       NaN       NaN  174.153807   65.882741   53.780203  103.147208   
std        NaN       NaN   12.431849    2.121442    2.457239   37.570528   
min        NaN       NaN  141.100000   60.300000   47.800000   48.000000   
25%        NaN       NaN  166.300000   64.100000   52.000000   70.000000   
50%        NaN       NaN  173.200000   65.500000   54.100000   95.000000   
75%        NaN       NaN  183.500000   66.900000   55.600000  116.000000   
max        NaN       NaN  208.100000   72.000000   59.800000  262.000000   

          city-mpg  highway-mpg         price  
count   197.000000   197.000000    197.000000  
unique         NaN          NaN           NaN  
top            NaN          NaN           NaN  
freq           NaN          NaN           NaN  
mean     25.238579    30.710660  13223.314721  
std       6.468837     6.875907   8015.753389  
min      13.000000    16.000000   5118.000000  
25%      19.000000    25.000000   7775.000000  
50%      25.000000    30.000000  10295.000000  
75%      30.000000    34.000000  16503.000000  
max      49.000000    54.000000  45400.000000

length, width, horsepower,price attributes contain NaN values at unique, top, freq

In [24]:
dataEmp = data2

In [25]:
dataEmp2 = pd.read_csv('data/car_country.csv')

In [28]:
cDataEmp = pd.merge(dataEmp, dataEmp2,left_on="make", right_on="car_maker")

In [38]:
cDataEmp

make fuel-type  length  width  height  horsepower  city-mpg  \
0    alfa-romero       gas   168.8   64.1    48.8       111.0        21   
1    alfa-romero       gas   168.8   64.1    48.8       111.0        21   
2           audi       gas   176.6   66.4    54.3       115.0        18   
3           audi       gas   177.3   66.3    53.1       110.0        19   
4           audi       gas   192.7   71.4    55.7       110.0        19   
..           ...       ...     ...    ...     ...         ...       ...   
192        volvo       gas   188.8   68.9    55.5       114.0        23   
193        volvo       gas   188.8   68.8    55.5       160.0        19   
194        volvo       gas   188.8   68.9    55.5       134.0        18   
195        volvo    diesel   188.8   68.9    55.5       106.0        26   
196        volvo       gas   188.8   68.9    55.5       114.0        19   

     highway-mpg    price    car_maker  country  
0             27  13495.0  alfa-romero    Italy  
1             27  16500.0  alfa-romero    Italy  
2             22  17450.0         audi  Germany  
3             25  15250.0         audi  Germany  
4             25  17710.0         audi  Germany  
..           ...      ...          ...      ...  
192           28  16845.0        volvo   Sweden  
193           25  19045.0        volvo   Sweden  
194           23  21485.0        volvo   Sweden  
195           27  22470.0        volvo   Sweden  
196           25  22625.0        volvo   Sweden  

[197 rows x 11 columns]

In [52]:
cDataEmp['country'].str.lower()

0        italy
1        italy
2      germany
3      germany
4      germany
        ...   
192     sweden
193     sweden
194     sweden
195     sweden
196     sweden
Name: country, Length: 197, dtype: object

In [30]:
first = ['country']

In [58]:
cDataEmp.groupby('country')['make'].count()

country
France     11
Germany    37
Italy       2
Japan      63
Sweden     17
UK         10
USA        13
japan      44
Name: make, dtype: int64

1-f

In [68]:
cDataEmp.groupby('country').aggregate({'price': min})

price
country         
France   11900.0
Germany   7775.0
Italy    13495.0
Japan     5195.0
Sweden   11850.0
UK        5572.0
USA       5151.0
japan     5118.0

In [69]:
cDataEmp.groupby('country').aggregate({'price': np.median})

price
country         
France   16630.0
Germany  20970.0
Italy    14997.5
Japan     8499.0
Sweden   16845.0
UK        8439.0
USA       6692.0
japan     8863.0

In [70]:
cDataEmp.groupby('country').aggregate({'price': max})

price
country         
France   18150.0
Germany  45400.0
Italy    16500.0
Japan    19699.0
Sweden   22625.0
UK       36000.0
USA      16503.0
japan    17669.0

In [71]:
from scipy.stats import ttest_1samp

In [73]:
H_mean = np.mean(cDataEmp['horsepower'])
H_mean

103.1472081218274

In [74]:
tset, pval = ttest_1samp(cDataEmp['horsepower'], 100)

In [75]:
print("p-values",pval)
if pval < 0.05: 
   print(" we are rejecting null hypothesis")
else:
  print("we are failing to reject null hypothesis")

p-values 0.24112496971220262
we are failing to reject null hypothesis


Null hypothesis: The mean of "horsepower" data equals to 100
Alternate hypothesis: The mean of "horsepower" data does not equal to 100
we are failing to reject null hypothesis, the mean of "horsepower" data equals to 100

1-h

In [76]:
import statsmodels.formula.api as smf

In [81]:
lm = smf.ols(formula='price ~ length', data=cDataEmp).fit()
lm.params

Intercept   -64800.967141
length         448.019387
dtype: float64

In [82]:
lm = smf.ols(formula='price ~ length', data=cDataEmp).fit()
lm.rsquared

0.4828101164752121

1-i

2-a

In [151]:
data4 = pd.read_csv('data/occupancy_data.csv', sep = ",")

In [152]:
data4.groupby('date').describe()

no                                                      \
                count    mean       std     min      25%     50%      75%   
date                                                                        
2/10/2015 00:00   1.0  7570.0       NaN  7570.0  7570.00  7570.0  7570.00   
2/10/2015 00:01   1.0  7571.0       NaN  7571.0  7571.00  7571.0  7571.00   
2/10/2015 00:02   1.0  7572.0       NaN  7572.0  7572.00  7572.0  7572.00   
2/10/2015 00:03   1.0  7573.0       NaN  7573.0  7573.00  7573.0  7573.00   
2/10/2015 00:04   2.0  7574.5  0.707107  7574.0  7574.25  7574.5  7574.75   
...               ...     ...       ...     ...      ...     ...      ...   
2/9/2015 23:53    1.0  7563.0       NaN  7563.0  7563.00  7563.0  7563.00   
2/9/2015 23:54    1.0  7564.0       NaN  7564.0  7564.00  7564.0  7564.00   
2/9/2015 23:55    2.0  7565.5  0.707107  7565.0  7565.25  7565.5  7565.75   
2/9/2015 23:57    2.0  7567.5  0.707107  7567.0  7567.25  7567.5  7567.75   
2/9/2015 23:58    1.0  7569.0       NaN  7569.0  7569.00  7569.0  7569.00   

                        Temperature             ... HumidityRatio            \
                    max       count       mean  ...           75%       max   
date                                            ...                           
2/10/2015 00:00  7570.0         1.0  20.390000  ...      0.004884  0.004884   
2/10/2015 00:01  7571.0         1.0  20.356667  ...      0.004881  0.004881   
2/10/2015 00:02  7572.0         1.0  20.390000  ...      0.004891  0.004891   
2/10/2015 00:03  7573.0         1.0  20.390000  ...      0.004886  0.004886   
2/10/2015 00:04  7575.0         2.0  20.377500  ...      0.004876  0.004877   
...                 ...         ...        ...  ...           ...       ...   
2/9/2015 23:53   7563.0         1.0  20.390000  ...      0.004877  0.004877   
2/9/2015 23:54   7564.0         1.0  20.390000  ...      0.004877  0.004877   
2/9/2015 23:55   7566.0         2.0  20.390000  ...      0.004877  0.004877   
2/9/2015 23:57   7568.0         2.0  20.390000  ...      0.004877  0.004877   
2/9/2015 23:58   7569.0         1.0  20.390000  ...      0.004891  0.004891   

                Occupancy                                     
                    count mean  std  min  25%  50%  75%  max  
date                                                          
2/10/2015 00:00       1.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
2/10/2015 00:01       1.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
2/10/2015 00:02       1.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
2/10/2015 00:03       1.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
2/10/2015 00:04       2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
...                   ...  ...  ...  ...  ...  ...  ...  ...  
2/9/2015 23:53        1.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
2/9/2015 23:54        1.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  
2/9/2015 23:55        2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2/9/2015 23:57        2.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2/9/2015 23:58        1.0  0.0  NaN  0.0  0.0  0.0  0.0  0.0  

[6514 rows x 56 columns]

2-b

In [128]:
data4.groupby('Occupancy').describe()

no                                                            \
            count         mean          std   min      25%     50%      75%   
Occupancy                                                                     
0          6414.0  4191.795447  2281.888540  17.0  2159.25  4348.5  5951.75   
1          1729.0  3627.599769  2542.430281   1.0  1332.00  2614.0  6813.00   

                  Temperature             ...      CO2         HumidityRatio  \
              max       count       mean  ...      75%     max         count   
Occupancy                                 ...                                  
0          8091.0      6414.0  20.334931  ...   464.50  1784.5        6414.0   
1          8143.0      1729.0  21.673192  ...  1131.25  2028.5        1729.0   

                                                                       \
               mean       std       min       25%       50%       75%   
Occupancy                                                               
0          0.003730  0.000753  0.002674  0.003004  0.003779  0.004274   
1          0.004355  0.001006  0.002715  0.003419  0.004347  0.005107   

                     
                max  
Occupancy            
0          0.006218  
1          0.006476  

[2 rows x 48 columns]

In [130]:
data4.groupby('Occupancy').aggregate({'Temperature':max})

Temperature
Occupancy             
0                23.10
1                23.18

In [132]:
data4.groupby('Occupancy').aggregate({'Temperature':min})

Temperature
Occupancy             
0                 19.0
1                 19.5

In [141]:
data4.groupby('Occupancy').aggregate({'Temperature':np.median})

Temperature
Occupancy             
0              20.2000
1              21.7675

If Occupancy is 0, max Temperature is 23.10, min Temperature is 19.0, median Temperature is 20.2000
If Occupancy is 1, max Temperature is 23.18, min Tempearture is 19.5, median Temperature is 21.7675
When Occupancy 1, Temperature is higher than 0 occupancy

In [138]:
data4.groupby('Occupancy').aggregate({'Humidity':max})

Humidity
Occupancy          
0           37.6900
1           39.1175

In [139]:
data4.groupby('Occupancy').aggregate({'Humidity':min})

Humidity
Occupancy          
0            16.745
1            18.600

In [142]:
data4.groupby('Occupancy').aggregate({'Humidity':np.median})

Humidity
Occupancy          
0           26.1875
1           26.4425

If Occupancy is 0, max Humidity is 37.69, min Humidity is 16.745, median Humidity is 26.1875
If Occupancy is 1, max Humidity is 39.1175, min Humidity is 18.6, median Humidity is 26.4425
When Occupancy 1, Humidity is higher than 0 occupancy

2-c

In [158]:
data4.resample('T').sum()

TypeError: Only valid with DatetimeIndex, TimedeltaIndex or PeriodIndex, but got an instance of 'RangeIndex'

2-d

In [157]:
data5 = data4.drop(['no','date','Light','CO2','HumidityRatio','Occupancy'], axis= 1)
data5

Temperature  Humidity
0           23.18   27.2720
1           23.15   27.2675
2           23.15   27.2450
3           23.15   27.2000
4           23.10   27.2000
...           ...       ...
8138        21.05   36.0975
8139        21.05   35.9950
8140        21.10   36.0950
8141        21.10   36.2600
8142        21.10   36.2000

[8143 rows x 2 columns]

2-e

In [144]:
from scipy.stats import ttest_ind
ttest,pval = ttest_ind(data4['Temperature'],data4['Humidity'])
print("ttest", ttest)
print("p-value",pval)

if pval <0.05:
  print("we reject null hypothesis")
else:
  print("we fail to reject (accept) null hypothesis")

ttest -82.03144157848514
p-value 0.0
we reject null hypothesis


Null hypothesis: The mean of "Temperature" data equals to the mean of "Humidity" data
Alternate hypothesis: The mean of "Temperature" data does not equal to the mean of "Humidity" data
We reject null hypothesis, So the mean of "Temperature" does not equal to the mean of "Humidity" data

2-f

In [154]:
by_time = data.groupby(data.index.time).mean()
hourly_ticks = 4 * 60 * 60 * np.arange(6)
by_time.plot(xticks=hourly_ticks, style=[':', '--', '-']);

AttributeError: 'RangeIndex' object has no attribute 'time'

2-g

In [155]:
by_weekday = data5.groupby(data.index.date).mean()
by_weekday.index = ['Mon', 'Tues', 'Wed', 'Thurs', 'Fri', 'Sat', 'Sun']
by_weekday.plot(style=[':', '--', '-']);

AttributeError: 'DataFrameGroupBy' object has no attribute 'groupby'